src/client/workspaceSymbols/main.ts

```diff
@@ -52,40 +61,53 @@ export class WorkspaceSymbols implements vscode.Disposable {
     dispose() {
         this.disposables.forEach(d => d.dispose());
     }
-    buildWorkspaceSymbols(rebuild: boolean = true, token?: vscode.CancellationToken): Promise<any> {
-        if (!pythonSettings.workspaceSymbols.enabled || (token && token.isCancellationRequested)) {
+    async buildWorkspaceSymbols(rebuild: boolean = true, token?: vscode.CancellationToken): Promise<any> {
+        if (token && token.isCancellationRequested) {
             return Promise.resolve([]);
         }
-        if (!vscode.workspace || typeof vscode.workspace.rootPath !== 'string' || vscode.workspace.rootPath.length === 0) {
+        if (this.generators.length === 0) {
             return Promise.resolve([]);
         }
-        return fsExistsAsync(pythonSettings.workspaceSymbols.tagFilePath).then(exits => {
-            let promise = Promise.resolve();
 
+        let promptPromise: Promise<InstallerResponse>;
+        let promptResponse: InstallerResponse;
+        return this.generators.map(async generator => {
+            if (!generator.enabled) {
+                return;
+            }
+            const exists = await fsExistsAsync(generator.tagFilePath);
             // if file doesn't exist, then run the ctag generator
             // Or check if required to rebuild
-            if (rebuild || !exits) {
-                promise = this.generator.generateWorkspaceTags();
+            if (!rebuild && exists) {
+                return;
             }
-
-            return promise.catch(reason => {
-                if (!isNotInstalledError(reason)) {
-                    this.outputChannel.show();
-                    return Promise.reject(reason);
+            for (let counter = 0; counter < MAX_NUMBER_OF_ATTEMPTS_TO_INSTALL_AND_BUILD; counter++) {
```

Comment:
This is just screaming `for ... in ...`, but I know there's no `range()` equivalent in TS without going to npm.